assets/src/edit-story/app/rightClickMenu/useRightClickMenu.js

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '..';
+import { noop } from '../../utils/noop';
+import { ELEMENT_TYPE } from '../highlights/quickActions/constants';
+import { RIGHT_CLICK_MENU_LABELS } from './constants';
+
+/**
+ * Determines the items displayed in the right click menu
+ * based off of the right-clicked element.
+ *
+ * Right click menu items should have the same shape as items
+ * in the design system's context menu.
+ *
+ * @return {Object} An object containing:
+ * - The items in the menu
+ * - The menu's position
+ * - Helper functions to open and close the menu
+ * - A ref for the container where the right click menu will be rendered
+ */
+const useRightClickMenu = () => {
+  const { selectedElements } = useStory(
+    ({
+      state: { currentPage, selectedElements, selectedElementAnimations },
+    }) => ({
+      currentPage,
+      selectedElements,
+      selectedElementAnimations,
+    })
+  );
+
+  // Ref to use to calculate the menu's position
+  const rightClickAreaRef = useRef();
+
+  const [isMenuOpen, setIsMenuOpen] = useState(false);
+  const [menuPosition, setMenuPosition] = useState({ x: 0, y: 0 });
+
+  /**
+   * Open the menu at the position from the click event.
+   */
+  const handleOpenMenu = useCallback((evt) => {
+    evt.preventDefault();
+    const layoutRect = rightClickAreaRef?.current?.getBoundingClientRect();
+
+    if (layoutRect) {
+      setMenuPosition({
+        x: evt.clientX - layoutRect?.left,
+        y: evt.clientY - layoutRect?.top,
+      });
```

Comment:
If the `contextmenu` handler is attached to an element that occupies the space you want this functionality to apply to, I believe just grabbing `offsetX` and `offsetY` basically does this math for you and saves you from a document style calc and stuff. Should allow the platform to do a bit more of the work for us